File: Codes/016_threeSumClosest/16.最接近的三数之和.go
/*
 * @lc app=leetcode.cn id=16 lang=golang
 *
 * [16] 最接近的三数之和
 */

// @lc code=start
func threeSumClosest(nums []int, target int) int {
	almost := math.MinInt64

	//0.排序
	sort.Ints(nums)
	//1.单遍历+双指针
	for i := 0; i < len(nums)-1; i++ {
		left := i + 1
		right := len(nums) - 1
		for {
			if left >= right {
				break
			}
			sum := nums[i] + nums[left] + nums[right]

			if sum > target {
				right--
			} else if sum == target {
				return target
			} else {
				left++
			}

			if math.Abs(float64(sum-target)) < math.Abs(float64(almost-target)) {
				almost = sum
			}
		}
	}

	return almost
}

// @lc code=end

File: Codes/016_threeSumClosest/16.最接近的三数之和.js
/*
 * @lc app=leetcode.cn id=16 lang=javascript
 *
 * [16] 最接近的三数之和
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    var almost=Infinity;

    nums.sort(function(a, b){return a - b}); 

    for (var i=0; i<nums.length-1; i++) {
        var left=i+1;
        var right=nums.length-1;
        while(left<right) {
            sum=nums[i]+nums[left]+nums[right];

            if(sum>target){
                right--;
            }else if(sum==target) {
                return target;
            }else{
                left++;
            }
            if(Math.abs(sum-target)<Math.abs(almost-target)){
                almost=sum;
            }
        }
    }

    return almost;

};
// @lc code=end

